overlord/assertstate/assertstate.go

```diff
@@ -889,10 +891,82 @@ func TryEnforceValidationSets(st *state.State, validationSets []string, userID i
 	return addCurrentTrackingToValidationSetsHistory(st)
 }
 
-// EnforceValidationSet tries to fetch the given validation set and enforce it.
+// ApplyEnforcedValidationSets enforces the supplied validation sets. It takes a map
+// of validation set keys to validation sets, pinned sequence numbers (if any),
+// installed snaps and ignored snaps. It fetches any pre-requisites necessary.
+func ApplyEnforcedValidationSets(st *state.State, valsets map[string]*asserts.ValidationSet, pinnedSeqs map[string]int, snaps []*snapasserts.InstalledSnap, ignoreValidation map[string]bool, userID int) error {
+	deviceCtx, err := snapstate.DevicePastSeeding(st, nil)
+	if err != nil {
+		return err
+	}
+
+	db := cachedDB(st)
+	batch := asserts.NewBatch(handleUnsupported(db))
+
+	vss := make([]*asserts.ValidationSet, 0, len(valsets))
+	valsetTracking := make([]*ValidationSetTracking, 0, len(valsets))
+
+	for vsKey, vs := range valsets {
```

Comment:
`vs` and `vss` are a bit confusing. IMHO `vs` is acceptable as long it's not used outside of the small loop where it's defined, but I would definitely rename vss at least, maybe to `valsetsSlice`.

Reply:
Thanks. I renamed that and `valsetTracking` to `valsetsTracking` to be more consistent